libraries/AP_BattMonitor/AP_BattMonitor_SMBus_Maxell.cpp

```diff
@@ -0,0 +1,94 @@
+#include <AP_Common/AP_Common.h>
+#include <AP_Math/AP_Math.h>
+#include "AP_BattMonitor.h"
+#include "AP_BattMonitor_SMBus_Maxell.h"
+#include <utility>
+
+extern const AP_HAL::HAL& hal;
+
+#include <AP_HAL/AP_HAL.h>
+
+#define BATTMONITOR_SMBUS_TEMP      0x08    // temperature register
+#define BATTMONITOR_SMBUS_VOLTAGE   0x09    // voltage register
+#define BATTMONITOR_SMBUS_CURRENT               0x0a    // current register
+#define BATTMONITOR_SMBUS_CHARGE_STATUS         0x0d    // relative state of charge
+#define BATTMONITOR_SMBUS_BATTERY_STATUS        0x16    // battery status register including alarms
+#define BATTMONITOR_SMBUS_BATTERY_CYCLE_COUNT   0x17    // cycle count
+#define BATTMONITOR_SMBUS_DESIGN_VOLTAGE        0x19    // design voltage register
+#define BATTMONITOR_SMBUS_SPECIFICATION_INFO    0x1a    // specification info
+#define BATTMONITOR_SMBUS_MANUFACTURE_NAME      0x1b    // manufacturer name
+#define BATTMONITOR_SMBUS_SERIALNUM             0x1c    // serial number register
+#define BATTMONITOR_SMBUS_CELL_VOLTAGE6         0x3a    // cell voltage register
+#define BATTMONITOR_SMBUS_CELL_VOLTAGE5         0x3b    // cell voltage register
+#define BATTMONITOR_SMBUS_CELL_VOLTAGE4         0x3c    // cell voltage register
+#define BATTMONITOR_SMBUS_CELL_VOLTAGE3         0x3d    // cell voltage register
+#define BATTMONITOR_SMBUS_CELL_VOLTAGE2         0x3e    // cell voltage register
+#define BATTMONITOR_SMBUS_CELL_VOLTAGE1         0x3f    // cell voltage register
+#define BATTMONITOR_SMBUS_HEALTH_STATUS         0x4f    // state of health
+#define BATTMONITOR_SMBUS_SAFETY_ALERT          0x50    // safety alert
+#define BATTMONITOR_SMBUS_SAFETY_STATUS         0x50    // safety status
+#define BATTMONITOR_SMBUS_PF_ALERT              0x52    // safety status
+#define BATTMONITOR_SMBUS_PF_STATUS             0x53    // safety status
+
+// Constructor
+AP_BattMonitor_SMBus_Maxell::AP_BattMonitor_SMBus_Maxell(AP_BattMonitor &mon, uint8_t instance,
+                                                   AP_BattMonitor::BattMonitor_State &mon_state,
+                                                   AP_HAL::OwnPtr<AP_HAL::I2CDevice> dev)
+    : AP_BattMonitor_SMBus(mon, instance, mon_state)
+    , _dev(std::move(dev))
+{
+    _dev->register_periodic_callback(100000, FUNCTOR_BIND_MEMBER(&AP_BattMonitor_SMBus_Maxell::timer, void));
+}
+
+/// Read the battery voltage and current.  Should be called at 10hz
+void AP_BattMonitor_SMBus_Maxell::read()
+{
+    // nothing to do - all done in timer()
+}
+
+void AP_BattMonitor_SMBus_Maxell::timer()
+{
+    uint16_t data;
+    uint32_t tnow = AP_HAL::micros();
+
+    // read voltage (V)
+    if (read_word(BATTMONITOR_SMBUS_VOLTAGE, data, 2)) {
+        _state.voltage = (float)data / 1000.0f;
+        _state.last_time_micros = tnow;
+        _state.healthy = true;
+    }
+
+    // timeout after 5 seconds
+    if ((tnow - _state.last_time_micros) > AP_BATTMONITOR_SMBUS_TIMEOUT_MICROS) {
```

Comment:
The battery overflows in 4295 seconds. I think that it is better to use the millis method because there is also a battery that lasts over an hour.